Jenkinsfile

```diff
@@ -33,69 +33,85 @@ def setRobotestParameters() {
-    stage('checkout') {
-      checkout scm
-      sh "git submodule update --init --recursive"
-      sh "sudo git clean -ffdx" // supply -f flag twice to force-remove untracked dirs with .git subdirs (e.g. submodules)
-    }
-    stage('params') {
-      // For try builds, we DO NOT want to overwrite parameters, as try builds
-      // offer a superset of PR/nightly parameters, and the extra ones will be
-      // lost when setRobotestParameters() is called -- 2020-04 walt
-      echo "Jenkins Job Parameters:"
-      for (param in params) { echo "${param}" }
-      if (env.KEEP_PARAMETERS == 'true') {
-        echo "KEEP_PARAMETERS detected. Ignoring Jenkins job parameters from Jenkinsfile."
-      } else {
-        echo "Overwriting Jenkins job parameters with parameters from Jenkinsfile."
-        setRobotestParameters()
-        propagateParamsToEnv()
-      }
+// robotest() defines the robotest pipeline.  It is expected to be run after
+// the gravity build, as it implicitly relies on artifacts & state from those targets.
+//
+// Per https://plugins.jenkins.io/pipeline-stage-view/: If we want to visualize
+// dynamically changing stages, it is better to make it conditional to execute the
+// stage contents, not conditional to include the stage.
+def robotest() {
+  stage('set robotest params') {
+    // For try builds, we DO NOT want to overwrite parameters, as try builds
+    // offer a superset of PR/nightly parameters, and the extra ones will be
+    // lost when setRobotestParameters() is called -- 2020-04 walt
+    echo "Jenkins Job Parameters:"
+    for (param in params) { echo "${param}" }
+    if (env.KEEP_PARAMETERS == 'true') {
+      echo "KEEP_PARAMETERS detected. Ignoring Jenkins job parameters from Jenkinsfile."
+    } else {
+      echo "Overwriting Jenkins job parameters with parameters from Jenkinsfile."
+      setRobotestParameters()
+      propagateParamsToEnv()
     }
-    stage('clean') {
-      sh "make -C e clean"
-    }
-    stage('build-gravity') {
-      withCredentials([
-      [$class: 'SSHUserPrivateKeyBinding', credentialsId: '08267d86-0b3a-4101-841e-0036bf780b11', keyFileVariable: 'GITHUB_SSH_KEY'],
-      [
-        $class: 'UsernamePasswordMultiBinding',
-        credentialsId: 'jenkins-aws-s3',
-        usernameVariable: 'AWS_ACCESS_KEY_ID',
-        passwordVariable: 'AWS_SECRET_ACCESS_KEY',
-      ],
-      ]) {
-        sh 'make -C e production telekube opscenter'
-      }
+  }
+  runRobotest = (env.RUN_ROBOTEST == 'run')
+  stage('build robotest images') {
+    if (runRobotest) {
+      sh 'make -C e/assets/robotest images'
+    } else {
+      echo 'skipping building robotest images'
     }
   }
   throttle(['robotest']) {
-    node {
-      stage('build-and-test') {
-        parallel (
-        build : {
+    stage('run robotest') {
+      if (runRobotest) {
+        withCredentials([
+					file(credentialsId:'ROBOTEST_LOG_GOOGLE_APPLICATION_CREDENTIALS', variable: 'GOOGLE_APPLICATION_CREDENTIALS'),
```

Comment:
Looks like use of spaces vs tabs is inconsistent in this file. Probably need to install a Groovy indent plugin.

Reply:
Fixed in f3a02f8.